File: generated-code/codecraft/go/TcpReadWrite/model/repair_properties.go
package model

import "io"
import . "trans_gen_test/stream"
import "fmt"

// Entity's repair properties
type RepairProperties struct {
    // Valid target entity types
    ValidTargets []EntityType
    // Health restored in one tick
    Power int32
}

func NewRepairProperties(validTargets []EntityType, power int32) RepairProperties {
    return RepairProperties {
        ValidTargets: validTargets,
        Power: power,
    }
}

// Read RepairProperties from reader
func ReadRepairProperties(reader io.Reader) RepairProperties {
    var validTargets []EntityType
    validTargets = make([]EntityType, ReadInt32(reader))
    for validTargetsIndex := range validTargets {
        var validTargetsElement EntityType
        validTargetsElement = ReadEntityType(reader)
        validTargets[validTargetsIndex] = validTargetsElement
    }
    var power int32
    power = ReadInt32(reader)
    return RepairProperties {
        ValidTargets: validTargets,
        Power: power,
    }
}

// Write RepairProperties to writer
func (repairProperties RepairProperties) Write(writer io.Writer) {
    validTargets := repairProperties.ValidTargets
    WriteInt32(writer, int32(len(validTargets)))
    for _, validTargetsElement := range validTargets {
        WriteInt32(writer, int32(validTargetsElement))
    }
    power := repairProperties.Power
    WriteInt32(writer, power)
}

// Get string representation of RepairProperties
func (repairProperties RepairProperties) String() string {
    stringResult := "{ "
    stringResult += "ValidTargets: "
    validTargets := repairProperties.ValidTargets
    stringResult += "[ "
    for validTargetsIndex, validTargetsElement := range validTargets {
        if validTargetsIndex != 0 {
            stringResult += ", "
        }
        stringResult += EntityTypeToString(validTargetsElement)
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "Power: "
    power := repairProperties.Power
    stringResult += fmt.Sprint(power)
    stringResult += " }"
    return stringResult
}

File: generated-code/csharp/codecraft/TcpReadWrite/Model/RepairProperties.cs
namespace TransGenTest.Model
{
    /// <summary>
    /// Entity's repair properties
    /// </summary>
    public struct RepairProperties
    {
        /// <summary>
        /// Valid target entity types
        /// </summary>
        public Model.EntityType[] ValidTargets { get; set; }
        /// <summary>
        /// Health restored in one tick
        /// </summary>
        public int Power { get; set; }
    
        public RepairProperties(Model.EntityType[] validTargets, int power)
        {
            this.ValidTargets = validTargets;
            this.Power = power;
        }
    
        /// <summary> Read RepairProperties from reader </summary>
        public static RepairProperties ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new RepairProperties();
            result.ValidTargets = new Model.EntityType[reader.ReadInt32()];
            for (int validTargetsIndex = 0; validTargetsIndex < result.ValidTargets.Length; validTargetsIndex++)
            {
                result.ValidTargets[validTargetsIndex] = EntityTypeHelper.ReadFrom(reader);
            }
            result.Power = reader.ReadInt32();
            return result;
        }
    
        /// <summary> Write RepairProperties to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(ValidTargets.Length);
            foreach (var validTargetsElement in ValidTargets)
            {
                writer.Write((int) (validTargetsElement));
            }
            writer.Write(Power);
        }
    
        /// <summary> Get string representation of RepairProperties </summary>
        public override string ToString() {
            string stringResult = "RepairProperties { ";
            stringResult += "ValidTargets: ";
            stringResult += "[ ";
            int validTargetsIndex = 0;
            foreach (var validTargetsElement in ValidTargets)
            {
                if (validTargetsIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += validTargetsElement.ToString();
                validTargetsIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "Power: ";
            stringResult += Power.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}